In [1]:
# Day 2 - Data Cleaning

import pandas as pd


In [2]:
nav = pd.read_csv("../data/raw/02_nav_history.csv")

print("Shape:", nav.shape)
print("\nColumns:")
print(nav.columns)

print("\nMissing Values:")
print(nav.isnull().sum())

print("\nDuplicate Rows:")
print(nav.duplicated().sum())

nav.head()

Shape: (46000, 3)

Columns:
Index(['amfi_code', 'date', 'nav'], dtype='str')

Missing Values:
amfi_code    0
date         0
nav          0
dtype: int64

Duplicate Rows:
0


,amfi_code,date,nav
0,119551,2022-01-03,54.3856
1,119551,2022-01-04,54.3474
2,119551,2022-01-05,54.6869
3,119551,2022-01-06,55.4550
4,119551,2022-01-07,55.3692


In [4]:
print(nav.dtypes)

amfi_code      int64
date             str
nav          float64
dtype: object


In [5]:
nav['date'] = pd.to_datetime(nav['date'])

print(nav.dtypes)

amfi_code             int64
date         datetime64[us]
nav                 float64
dtype: object


In [6]:
#checking NAV values
print("Minimum NAV:", nav['nav'].min())
print("Maximum NAV:", nav['nav'].max())

Minimum NAV: 26.1366
Maximum NAV: 4268.5497


In [7]:
# Sort dataset
nav = nav.sort_values(
    ['amfi_code', 'date']
)

# Save cleaned file
nav.to_csv(
    "../data/processed/clean_nav_history.csv",
    index=False
)

print("clean_nav_history.csv saved successfully")

clean_nav_history.csv saved successfully


In [8]:
import os

print(
    os.path.exists(
        "../data/processed/clean_nav_history.csv"
    )
)

True


## Cleaning Summary - 02_nav_history.csv

- Rows: 46,000
- Missing Values: 0
- Duplicate Rows: 0
- Date converted to datetime
- NAV validated (>0)
- Dataset sorted by amfi_code and date
- Saved as clean_nav_history.csv

In [10]:
transactions = pd.read_csv(
    "../data/raw/08_investor_transactions.csv"
)

print("Shape:", transactions.shape)

print("\nColumns:")
print(transactions.columns)

print("\nMissing Values:")
print(transactions.isnull().sum())

print("\nDuplicate Rows:")
print(transactions.duplicated().sum())

transactions.head()

Shape: (32778, 13)

Columns:
Index(['investor_id', 'transaction_date', 'amfi_code', 'transaction_type',
       'amount_inr', 'state', 'city', 'city_tier', 'age_group', 'gender',
       'annual_income_lakh', 'payment_mode', 'kyc_status'],
      dtype='str')

Missing Values:
investor_id           0
transaction_date      0
amfi_code             0
transaction_type      0
amount_inr            0
state                 0
city                  0
city_tier             0
age_group             0
gender                0
annual_income_lakh    0
payment_mode          0
kyc_status            0
dtype: int64

Duplicate Rows:
0


,investor_id,transaction_date,amfi_code,transaction_type,amount_inr,state,city,city_tier,age_group,gender,annual_income_lakh,payment_mode,kyc_status
0,INV003054,2024-01-01,119092,SIP,1834,Telangana,Hyderabad,T30,56+,Female,77.1,UPI,Verified
1,INV002952,2024-01-01,148567,Redemption,392882,Punjab,Amritsar,B30,18-25,Male,7.1,Cheque,Verified
2,INV003420,2024-01-01,118636,SIP,912,Haryana,Faridabad,B30,36-45,Male,47.2,Mandate,Verified
3,INV003436,2024-01-01,118634,SIP,1102,Maharashtra,Mumbai,T30,36-45,Female,54.4,Cheque,Pending
4,INV004691,2024-01-01,119094,Lumpsum,8682,Delhi,Noida,T30,26-35,Male,14.5,Net Banking,Pending


In [11]:
print(transactions['transaction_type'].unique())

<ArrowStringArray>
['SIP', 'Redemption', 'Lumpsum']
Length: 3, dtype: str


In [12]:
print(transactions['kyc_status'].unique())

<ArrowStringArray>
['Verified', 'Pending']
Length: 2, dtype: str


In [13]:
print("Minimum Amount:", transactions['amount_inr'].min())
print("Maximum Amount:", transactions['amount_inr'].max())

Minimum Amount: 400
Maximum Amount: 597498


In [14]:
print(transactions.dtypes)

investor_id               str
transaction_date          str
amfi_code               int64
transaction_type          str
amount_inr              int64
state                     str
city                      str
city_tier                 str
age_group                 str
gender                    str
annual_income_lakh    float64
payment_mode              str
kyc_status                str
dtype: object


In [15]:
transactions['transaction_date'] = pd.to_datetime(
    transactions['transaction_date']
)

In [16]:
transactions.to_csv(
    "../data/processed/clean_investor_transactions.csv",
    index=False
)

print("clean_investor_transactions.csv saved")

clean_investor_transactions.csv saved


In [17]:
print("Transaction Types:")
print(transactions['transaction_type'].unique())

print("\nKYC Status:")
print(transactions['kyc_status'].unique())

print("\nMinimum Amount:")
print(transactions['amount_inr'].min())

Transaction Types:
<ArrowStringArray>
['SIP', 'Redemption', 'Lumpsum']
Length: 3, dtype: str

KYC Status:
<ArrowStringArray>
['Verified', 'Pending']
Length: 2, dtype: str

Minimum Amount:
400


## Cleaning Summary - 08_investor_transactions.csv

- Rows: 32,778
- Missing Values: 0
- Duplicate Rows: 0
- Transaction Types: SIP, Redemption, Lumpsum
- KYC Status: Verified, Pending
- Minimum Amount: 400
- transaction_date converted to datetime
- Saved as clean_investor_transactions.csv

In [18]:
performance = pd.read_csv(
    "../data/raw/07_scheme_performance.csv"
)

print("Shape:", performance.shape)

print("\nColumns:")
print(performance.columns)

print("\nMissing Values:")
print(performance.isnull().sum())

print("\nDuplicate Rows:")
print(performance.duplicated().sum())

performance.head()

Shape: (40, 19)

Columns:
Index(['amfi_code', 'scheme_name', 'fund_house', 'category', 'plan',
       'return_1yr_pct', 'return_3yr_pct', 'return_5yr_pct',
       'benchmark_3yr_pct', 'alpha', 'beta', 'sharpe_ratio', 'sortino_ratio',
       'std_dev_ann_pct', 'max_drawdown_pct', 'aum_crore', 'expense_ratio_pct',
       'morningstar_rating', 'risk_grade'],
      dtype='str')

Missing Values:
amfi_code             0
scheme_name           0
fund_house            0
category              0
plan                  0
return_1yr_pct        0
return_3yr_pct        0
return_5yr_pct        0
benchmark_3yr_pct     0
alpha                 0
beta                  0
sharpe_ratio          0
sortino_ratio         0
std_dev_ann_pct       0
max_drawdown_pct      0
aum_crore             0
expense_ratio_pct     0
morningstar_rating    0
risk_grade            0
dtype: int64

Duplicate Rows:
0


,amfi_code,scheme_name,fund_house,category,plan,return_1yr_pct,return_3yr_pct,return_5yr_pct,benchmark_3yr_pct,alpha,beta,sharpe_ratio,sortino_ratio,std_dev_ann_pct,max_drawdown_pct,aum_crore,expense_ratio_pct,morningstar_rating,risk_grade
0,119551,SBI Bluechip Fund - Regular Plan - Growth,SBI Mutual Fund,Large Cap,Regular,12.42,12.36,14.45,11.49,0.87,0.89,0.88,1.29,14.0,-21.70,14288,1.54,4,Moderate
1,119552,SBI Bluechip Fund - Direct Plan - Growth,SBI Mutual Fund,Large Cap,Direct,15.25,11.30,14.23,9.52,1.78,0.87,0.81,1.29,14.0,-24.43,1231,0.66,3,Moderate
2,119598,SBI Small Cap Fund - Regular Plan - Growth,SBI Mutual Fund,Small Cap,Regular,24.56,23.39,20.67,22.16,1.23,0.89,0.94,1.35,25.0,-13.35,19259,1.43,5,Very High
3,119599,SBI Small Cap Fund - Direct Plan - Growth,SBI Mutual Fund,Small Cap,Direct,20.59,23.14,21.82,22.01,1.13,1.04,0.93,1.67,25.0,-24.78,36061,0.72,4,Very High
4,119120,SBI Magnum Gilt Fund - Regular Plan - Growth,SBI Mutual Fund,Gilt,Regular,5.34,6.07,5.43,4.47,1.60,0.22,1.52,2.11,4.0,-2.30,24101,0.77,5,Low


In [19]:
nav = pd.read_csv("../data/raw/02_nav_history.csv")

print("Shape:", nav.shape)
print("\nColumns:")
print(nav.columns)

print("\nMissing Values:")
print(nav.isnull().sum())

print("\nDuplicate Rows:")
print(nav.duplicated().sum())

nav.head()

Shape: (46000, 3)

Columns:
Index(['amfi_code', 'date', 'nav'], dtype='str')

Missing Values:
amfi_code    0
date         0
nav          0
dtype: int64

Duplicate Rows:
0


,amfi_code,date,nav
0,119551,2022-01-03,54.3856
1,119551,2022-01-04,54.3474
2,119551,2022-01-05,54.6869
3,119551,2022-01-06,55.4550
4,119551,2022-01-07,55.3692


In [20]:
print(performance.dtypes)

amfi_code               int64
scheme_name               str
fund_house                str
category                  str
plan                      str
return_1yr_pct        float64
return_3yr_pct        float64
return_5yr_pct        float64
benchmark_3yr_pct     float64
alpha                 float64
beta                  float64
sharpe_ratio          float64
sortino_ratio         float64
std_dev_ann_pct       float64
max_drawdown_pct      float64
aum_crore               int64
expense_ratio_pct     float64
morningstar_rating      int64
risk_grade                str
dtype: object


In [22]:
print("Expense Ratio Range")
print("Minimum:", performance['expense_ratio_pct'].min())
print("Maximum:", performance['expense_ratio_pct'].max())

print("\nNegative Sharpe Ratios")
print((performance['sharpe_ratio'] < 0).sum())

print("\nReturn Statistics")
print(
    performance[
        ['return_1yr_pct',
        'return_3yr_pct',
        'return_5yr_pct']
    ].describe()
)

Expense Ratio Range
Minimum: 0.55
Maximum: 1.64

Negative Sharpe Ratios
0

Return Statistics
       return_1yr_pct  return_3yr_pct  return_5yr_pct
count       40.000000       40.000000       40.000000
mean        14.376000       14.089000       14.516750
std          4.883023        4.617253        4.454021
min          4.260000        5.140000        5.430000
25%         11.735000       12.035000       12.340000
50%         14.620000       14.205000       14.185000
75%         16.392500       15.882500       17.585000
max         24.930000       23.390000       23.800000


In [23]:
performance.to_csv(
    "../data/processed/clean_scheme_performance.csv",
    index=False
)

print("clean_scheme_performance.csv saved")

clean_scheme_performance.csv saved


In [24]:
performance.to_csv(
    "../data/processed/clean_scheme_performance.csv",
    index=False
)

print("Saved")

Saved


In [25]:
import os

print(
    os.listdir("../data/processed")
)

['.gitkeep', 'clean_investor_transactions.csv', 'clean_nav_history.csv', 'clean_scheme_performance.csv']


In [26]:
from sqlalchemy import create_engine

print("SQLAlchemy working")

SQLAlchemy working


In [27]:
engine = create_engine(
    "sqlite:///../data/bluestock_mf.db"
)

print("Database Created Successfully")

Database Created Successfully


In [28]:
clean_nav = pd.read_csv(
    "../data/processed/clean_nav_history.csv"
)

clean_nav.to_sql(
    "fact_nav",
    engine,
    if_exists="replace",
    index=False
)

print("fact_nav loaded")

fact_nav loaded


In [29]:
clean_transactions = pd.read_csv(
    "../data/processed/clean_investor_transactions.csv"
)

clean_transactions.to_sql(
    "fact_transactions",
    engine,
    if_exists="replace",
    index=False
)

print("fact_transactions loaded")

fact_transactions loaded


In [30]:
clean_performance = pd.read_csv(
    "../data/processed/clean_scheme_performance.csv"
)

clean_performance.to_sql(
    "fact_performance",
    engine,
    if_exists="replace",
    index=False
)

print("fact_performance loaded")

fact_performance loaded


In [31]:
import sqlite3

conn = sqlite3.connect(
    "../data/bluestock_mf.db"
)

cursor = conn.cursor()

cursor.execute(
    "SELECT name FROM sqlite_master WHERE type='table';"
)

print(cursor.fetchall())

conn.close()

[('fact_nav',), ('fact_transactions',), ('fact_performance',)]


In [32]:
conn = sqlite3.connect(
    "../data/bluestock_mf.db"
)

cursor = conn.cursor()

tables = [
    "fact_nav",
    "fact_transactions",
    "fact_performance"
]

for table in tables:
    cursor.execute(
        f"SELECT COUNT(*) FROM {table}"
    )
    count = cursor.fetchone()[0]
    print(f"{table}: {count}")

conn.close()

fact_nav: 46000
fact_transactions: 32778
fact_performance: 40


In [33]:
clean_performance.columns

Index(['amfi_code', 'scheme_name', 'fund_house', 'category', 'plan',
       'return_1yr_pct', 'return_3yr_pct', 'return_5yr_pct',
       'benchmark_3yr_pct', 'alpha', 'beta', 'sharpe_ratio', 'sortino_ratio',
       'std_dev_ann_pct', 'max_drawdown_pct', 'aum_crore', 'expense_ratio_pct',
       'morningstar_rating', 'risk_grade'],
      dtype='str')

In [34]:
conn = sqlite3.connect("../data/bluestock_mf.db")

cursor = conn.cursor()

tables = [
    "fact_nav",
    "fact_transactions",
    "fact_performance"
]

for table in tables:
    cursor.execute(
        f"SELECT COUNT(*) FROM {table}"
    )
    print(f"{table}: {cursor.fetchone()[0]}")

conn.close()

fact_nav: 46000
fact_transactions: 32778
fact_performance: 40
